chrome-extension/sidebar/sidebar.js

```diff
@@ -48,6 +48,290 @@ const htmlToElement = (html) => {
+};
+
+const sortBySeatsAvailable = () => {
+  // Looks like { "CSCI4011":26, "CSCI4041":37 ... }
+  const courseSizeDict = {};
+
+  // Get the main container with class 'course-list-results'
+  let courseListResults = document.querySelector('.course-list-results');
+  let courseDivs = Array.from(courseListResults.firstElementChild.children);
+
+  for (let courseDiv of courseDivs) {
+    // Get name of course
+    let courseName = courseDiv.querySelector('a').getAttribute('name');
+    
+    // Get seats available value
+    let targetDiv = courseDiv.querySelector('div:nth-child(2) > div:nth-child(3)');
+    let trElements = targetDiv.querySelectorAll('tr');
+    
+    let sum = 0;
+    trElements.forEach(tr => {
+      let lastChildText = tr.lastElementChild.innerText;
+      sum += parseFloat(extractDifference(lastChildText)) || 0; // Ensure the text is converted to a number
+    });
+    
+    courseSizeDict[courseName] = sum;
+  }
+
+  // sort the courseDivs in the order we want
+  courseDivs.sort((a, b) => {
+    let nameA = a.querySelector('a').getAttribute('name');
+    let nameB = b.querySelector('a').getAttribute('name');
+    return courseSizeDict[nameB] - courseSizeDict[nameA];
+  });
+
+  console.log("Sorted Seats Dict: ", courseSizeDict); // For debugging purposes, display the updated dictionary
+  console.log("Sorted Seats Course Divs: ", courseDivs); // For debugging purposes, display the updated divs
+
+  // remove all the courses to add our own ordering of the courses
+  courseListResults.firstElementChild.innerHTML = '';
+  courseDivs.forEach(courseDiv => {
+    courseListResults.firstElementChild.appendChild(courseDiv);
+  });
+
+  // click all the 'hide' buttons
+  const buttons2 = document.querySelectorAll('.action-sections.btn.btn-default.pull-right');
+  buttons2.forEach(button => button.click());
+
+  // ensure this value is selected
+  document.querySelector('.size-dropdown').value = 'sortSeatsAvailDes';
+};
+
+// Helper function to determine seats available
+function extractDifference(text) {
+  let matches = text.match(/(\d+)\s+of\s+(\d+)/g);
+  let totalDifference = 0;
+  
+  if (matches) {
+    matches.forEach(match => {
+      let parts = match.match(/(\d+)\s+of\s+(\d+)/);
+      let current = parseInt(parts[1]);
+      let total = parseInt(parts[2]);
+      totalDifference += (total - current);
+    });
+  }
+  return totalDifference;
+}
+
+const sortByMostCommonGrade = () => {
+  let courseMostCommonGradeDict = {};
+
+  // Get the main container with class 'course-list-results'
+  let courseListResults = document.querySelector('.course-list-results');
+  let courseDivs = Array.from(courseListResults.firstElementChild.children);
+
+  // Array to hold all fetch promises
+  let fetchPromises = [];
+
+  for (let courseDiv of courseDivs) {
+    let courseName = courseDiv.querySelector('a').getAttribute('name');
+
+    // Push the fetch promise into the array
+    let fetchPromise = new Promise((resolve, reject) => {
```

Comment:
we should load each class only once, instead of triggering a network call every time i want to sort by something else. since someone who sorts once is the most likely person to sort a second time, we should optimize for this from a UX standpoint.

Reply:
again, this might be easier to do after abstracting out